working/0723-static-extensions/feature-specification.md

```diff
@@ -1,35 +1,88 @@
+generalization of the features offered by extension declarations.
 
 ## Introduction
 
-A feature like extensions with static capabilities was requested already
-several years ago in [language issue #723][issue 723], and elsewhere.
+An extension declaration in Dart can already (before the addition of
+the feature specified here) declare static members, but not
+constructors.  Members so declared are only accessible via prefixing
+with the extension name.
+
+```dart
+extension Numbers on int {
+  static int one = 1;
+  static int two = 2;
+}
+void main() {
+  // Static members declared in an extension can be accessed by prefixing
+  // with the extension name.
+  print(Numbers.one);
+  print(Numbers.two);
+}
+```
+
+Developers have requested (e.g. [language issue #723][issue 723]) the
+capability of using extension declarations to declare static members
+(including constructors) which are accessible via the name of an
+existing class, mixin, enum, or extension type declaration.
 
 [issue 723]: https://github.com/dart-lang/language/issues/723
 
-The main motivation for this feature is that developers wish to add
-constructors or static members to an existing class, mixin, enum, or
-extension type declaration, but they do not have the ability to directly
-edit the source code of said declaration.
 
-This feature allows static members and constructors declared in an
-`extension` on a given class/mixin/etc. declaration _D_ to be invoked as if
-they were static members respectively constructors declared by _D_.
+This would allow static members and constructors to be added to
+existing declarations after the fact, including in situations where
+the user does not have ability to directly edit the source code of
+said declaration.
+
+The feature proposed here allows static members and constructors
+declared in an extension on a given class/mixin/etc. declaration _D_
+to be invoked as if they were static members (respectively
+constructors) declared by _D_.  In the case where the on-type of an
+extension declaration satisfies certain constraints defined below, we
+say that the class/mixin/etc. which is referred to in the on-type is
+the _on-declaration_ of the extension, and in the case that an
+extension has an on-declaration, static members and constructors
+declared in the extension become accessible via its on-declaration
+name.  So for example, in the extension `Numbers` defined above, the
+on-declaration of the extension is `int`, and the static members
+declared in the extension become accessible as if they were static
+members declared on the `int` class.
+
+```dart
+void main() {
+   // Static members on an extension are available via the on-declaration
+   // name
+   print(int.one + int.two);
+
+   // Static members on an extension continue to be available via the 
+   // extension name.
+   print(Numbers.one + Numbers.two);
+}
+```
+
+The enhancements specified for extension declarations in this
```

Comment:
This is surprising to me. Aren't all extensions required to have an on-declaration?

Reply:
No (see reference above to definition below).  on-declaration != on-type